components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp

```diff
@@ -0,0 +1,130 @@
+#ifndef SHOC_COMPUTE_DIAG_THIRD_SHOC_MOMENT_IMPL_HPP
+#define SHOC_COMPUTE_DIAG_THIRD_SHOC_MOMENT_IMPL_HPP
+
+#include "shoc_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace shoc {
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::compute_diag_third_shoc_moment(
+  const MemberType& team,
+  const Int& nlev,
+  const Int& nlevi,
+  const uview_1d<const Spack>& w_sec,
+  const uview_1d<const Spack>& thl_sec,
+  const uview_1d<const Spack>& wthl_sec,
+  const uview_1d<const Spack>& tke,
+  const uview_1d<const Spack>& dz_zt,
+  const uview_1d<const Spack>& dz_zi,
+  const uview_1d<const Spack>& isotropy_zi,
+  const uview_1d<const Spack>& brunt_zi,
+  const uview_1d<const Spack>& w_sec_zi,
+  const uview_1d<const Spack>& thetal_zi,
+  const uview_1d<Spack>& w3)
+{
+  const Int nlev_pack = ekat::pack::npack<Spack>(nlev);
+  const Int nlevi_pack = ekat::pack::npack<Spack>(nlevi);
+
+  // Scalarize views for shifts
+  const auto s_dz_zt = scalarize(dz_zt);
+  const auto s_wthl_sec = scalarize(wthl_sec);
+  const auto s_thl_sec = scalarize(thl_sec);
+  const auto s_w_sec = scalarize(w_sec);
+  const auto s_tke = scalarize(tke);
+
+  // Set lower condition: w3(i,nlevi) = 0
+  const Int last_pack_entry = (nlevi%Spack::n == 0 ? Spack::n-1 : nlevi%Spack::n-1);
+  w3(nlevi_pack-1)[last_pack_entry] = 0.0;
+
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team, nlev_pack), [&] (const Int& k) {
+    // Local variables
+    Spack
+      thedz, thedz2, iso, isosqrd, buoy_sgs2, bet2,
+      f0, f1, f2, f3, f4, f5,
+      omega0, omega1, omega2,
+      x0, x1, y0, y1,
+      aa0, aa1;
+
+    // Constants
+    const auto c_diag_3rd_mom = scream::shoc::Constants<Scalar>::c_diag_3rd_mom;
+    const Scalar a0 = (sp(0.52)*std::pow(c_diag_3rd_mom,-2.0))/(c_diag_3rd_mom-2.0);
+    const Scalar a1 = sp(0.87)/(c_diag_3rd_mom*c_diag_3rd_mom);
+    const Scalar a2 = sp(0.5)/c_diag_3rd_mom;
+    const Scalar a3 = sp(0.6)/(c_diag_3rd_mom*(c_diag_3rd_mom-2.0));
+    const Scalar a4 = sp(2.4)/(3.0*c_diag_3rd_mom+5.0);
+    const Scalar a5 = sp(0.6)/(c_diag_3rd_mom*(3.0+5.0*c_diag_3rd_mom));
+
+    // Calculate shifts
+    Spack
+      dz_zt_k,    dz_zt_km1,
+      wthl_sec_k, wthl_sec_km1, wthl_sec_kp1,
+      thl_sec_k,  thl_sec_km1,  thl_sec_kp1,
+      w_sec_k,    w_sec_km1,
+      tke_k,      tke_km1;
+
+    auto range_pack1 = ekat::pack::range<IntSmallPack>(k*Spack::n);
+    auto range_pack2 = range_pack1;
+    // index for _km1 should never go below 0 and _kp1 should never go above nlevi
+    range_pack2.set(range_pack1 < 1 || range_pack1 >= nlevi, 1);
```

Comment:
Is masking `range_pack1 < 1` right even when the shift is +1? Seems to me the 0 -> 1 index shift will be wrong.

Reply:
The loop is `k = 2,...,nlev` in shoc.F90, (switching to c++ 0 based units) so we never set actually use the 0->1 shift.